pkg/utils/set/set.go

```diff
@@ -0,0 +1,76 @@
+package set
+
+import (
+	"sync"
+)
+
+// Set is a generic thread-safe data structure representing a unique collection of items
+type Set[T comparable] struct {
+	items     []T
+	uniqueSet map[T]struct{}
+	mutex     *sync.RWMutex
+}
+
+func New[T comparable](items ...T) *Set[T] {
+	set := Set[T]{
+		items:     make([]T, 0, len(items)),
+		uniqueSet: make(map[T]struct{}, len(items)),
+		mutex:     new(sync.RWMutex),
+	}
+	for _, item := range items {
+		set.uniqueSet[item] = struct{}{}
+	}
+	for item := range set.uniqueSet {
+		set.items = append(set.items, item)
+	}
+	return &set
+}
+
+func (s *Set[T]) Empty() bool {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	return len(s.items) == 0
+}
+
+func (s *Set[T]) Has(item T) bool {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	_, ok := s.uniqueSet[item]
+	return ok
+}
+
+func (s *Set[T]) Items() []T {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	return s.items
```

Comment:
Good idea.

Reply:
I dont think we should have unsafe methods. Full encapsulation only. And I think we should enforce that pattern from now on. One of our biggest problems for refactoring is/was lack of encapsulation. IF a caller needs the "internal" data, either register a callback pointer to a function, deal with a channel for readiness completion OR be part of the object. Is that ok for you ?

Reply:
Not sure why it doesn't reflect here, in the current version `Items` returns a copy and the original was renamed `ItemsUnsafe`. I think that as reviewers we know to be extra careful if we see a code section using the Unsafe methods (which are unused in this code). IMO our issue so far has not been only lack of encapsulation but lack of safe encapsulation and unawareness it was unsafe. The code here is neither of these things, it is correctly encapsulated and unsafety is explicitly mentioned. 
I can remove it anyway, but I think if we use the Set in more critical sections it could be useful for reads. Using a callback instead does not guarantee any read safety.

Reply:
I've found a usage for the unsafe versions in the internal methods of the dnscache. I can see the criticism of this being a premature optimization, but i do think it's better as is. The name indicates that any code change near it should be done with care.